buzhash.go

```diff
@@ -0,0 +1,115 @@
+package chunk
+
+import (
+	"io"
+	"math/bits"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+const (
+	buzMin  = 128 << 10
+	buzMax  = 512 << 10
+	buzMask = 1<<17 - 1
+)
+
+type Buzhash struct {
+	r   io.Reader
+	buf []byte
+	n   int
+
+	err error
+}
+
+func NewBuzhash(r io.Reader) *Buzhash {
+	return &Buzhash{
+		r:   r,
+		buf: pool.Get(buzMax),
+	}
+}
+
+func (b *Buzhash) NextBytes() ([]byte, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	buf := b.buf
+	n, err := io.ReadFull(b.r, buf[b.n:])
+	if err != nil {
+		if err == io.ErrUnexpectedEOF {
+			b.err = io.EOF
+			return buf[:n+b.n], nil
+		} else {
+			b.err = err
+			pool.Put(buf)
+			return nil, err
+		}
+	}
+
+	i := buzMin - 32
+
+	var state uint32 = 0
+
+	for ; i < buzMin; i++ {
+		state = bits.RotateLeft32(state, 1)
+		state = state ^ bytehash[buf[i]]
+	}
+
+	for ; state&buzMask != 0; i++ {
+		if i >= buzMax {
+			break
+		}
+		state = bits.RotateLeft32(state, 1) ^ bytehash[buf[i-32]] ^ bytehash[buf[i]]
+	}
+
+	res := buf[:i]
+	b.buf = pool.Get(buzMax)
+	b.n = copy(b.buf, buf[i:])
```

Comment:
Instead of returning a poll-allocated (over allocated) buffer, we should _keep_ the pool buffer and instead allocate a new exact-sized buffer to return. Otherwise, we'll end up passing around (and _caching_ :scream:) blocks with over-allocated buffers.

(we've been here before...)

Reply:
The over-allocation is expected to be 25% on average. Is it worst to use an over-allocated buffer with 25% overhead, or allocate repeatedly?

Reply:
* It'll be 2x on average, unless I'm misreading something (512 = 2 * 256)
* We're allocating _anyways_ as we're not putting anything _into_ the pool. Actually, using a pool like this will probably allocate more (although I guess it can steal from other services).

Reply:
25% was my mistake.

Should we also remove the pool from fixed chunker?

Reply:
Changes when not using pool:
```
name       old time/op    new time/op     delta
Buzhash-4    14.2ms ± 7%     19.2ms ± 9%      +35.47%  (p=0.000 n=20+20)

name       old speed      new speed       delta
Buzhash-4  1.18GB/s ± 6%   0.88GB/s ± 8%      -26.13%  (p=0.000 n=20+20)

name       old alloc/op   new alloc/op    delta
Buzhash-4    13.8kB ±40%  17554.4kB ± 0%  +127447.66%  (p=0.000 n=19+20)

name       old allocs/op  new allocs/op   delta
Buzhash-4      1.00 ± 0%      67.50 ± 7%    +6650.00%  (p=0.000 n=20+20)
```


If none of them use pool:
```
name       time/op
Buzhash-4    18.9ms ± 5%
Rabin-4      95.3ms ± 7%
Default-4    4.26ms ± 2%

name       speed
Buzhash-4   889MB/s ± 5%
Rabin-4     176MB/s ± 6%
Default-4  3.93GB/s ± 2%

name       alloc/op
Buzhash-4    17.6MB ± 0%
Rabin-4      19.2MB ± 0%
Default-4    17.0MB ± 0%

name       allocs/op
Buzhash-4      66.6 ±10%
Rabin-4         198 ±10%
Default-4      67.0 ± 0%
```